kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateCloudRepository.java

```diff
@@ -0,0 +1,164 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package org.kie.server.services.openshift.impl.storage.cloud;
+
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.validation.constraints.NotNull;
+
+import com.thoughtworks.xstream.XStream;
+import com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider;
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import org.kie.server.api.KieServerConstants;
+import org.kie.server.api.model.KieContainerResource;
+import org.kie.server.api.model.KieServerConfigItem;
+import org.kie.server.services.impl.KieServerLocator;
+import org.kie.server.services.impl.storage.KieServerState;
+import org.kie.server.services.impl.storage.KieServerStateRepository;
+import org.kie.server.services.openshift.api.KieServerReadinessProbe;
+import org.kie.soup.commons.xstream.XStreamUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KieServerStateCloudRepository implements KieServerStateRepository,
+                                           KieServerReadinessProbe, CloudClientFactory {
+
+    public static final String ROLLOUT_REQUIRED = "services.server.kie.org/openshift-startup-strategy.rolloutRequired";
+    public static final String CFG_MAP_DATA_KEY = "kie-server-state";
+    private static final String STATE_CHANGE_TIMESTAMP = "services.server.kie.org/kie-server-state.changeTimestamp";
+    private static final Logger logger = LoggerFactory.getLogger(KieServerStateCloudRepository.class);
+
+    private final XStream xs;
+
+    public static XStream initializeXStream() {
+        XStream xs = XStreamUtils.createTrustingXStream(new PureJavaReflectionProvider());
+        String[] voidDeny = {"void.class", "Void.class"};
+        xs.denyTypes(voidDeny);
+        xs.alias(CFG_MAP_DATA_KEY, KieServerState.class);
+        xs.alias("container", KieContainerResource.class);
+        xs.alias("config-item", KieServerConfigItem.class);
+
+        return xs;
+    }
+
+    public KieServerStateCloudRepository() {
+        xs = initializeXStream();
+    }
+
+    public synchronized void create(@NotNull KieServerState kieServerState) {
+        createOrUpdate(retrieveKieServerId(kieServerState), kieServerState, false);
+    }
+
+    @Override
+    public synchronized void store(@NotNull String serverId, @NotNull KieServerState kieServerState) {
+        if (!retrieveKieServerId(kieServerState).equals(serverId)) {
+            throw new IllegalArgumentException("Invalid KieServerId: Id does not match with KieServerState.");
+        }
+        createOrUpdate(serverId, kieServerState, true);
```

Comment:
just thinking at loud ... do we really need to store it back to config map? I assume that the whole point of OpenShift based approach is to rely on external configuration that should not be modified by Kie Server itself, isn't it?

So maybe we should always load from config map but store it only on local (pods) file system or even ignore the store part completely... @rhtevan @sutaakar wdyt guys?

Reply:
Good question.
From what I understood there are 2 main goals which are intended to be provided by this implementation:
1. Provide alternative source of Kie server state info backed by config map (so users don't have to provide state file for Kie server pods).
2. Add possibility to scale unmanaged Kie server pods while keeping their state in sync (by respinning pods in case of any modification of Kie server state for specific Kie server id).

Reply:
Let's assume this scenario. Starting with a 'MyKieServer' with two Pods (i.e. MyKieServer-Pod-X, MyKieServer-Pod-Y) running for it without any pre-configured KieContainers.

1. A user issues a 'PUT /server/containers/{id}' call to deploy a new KieContainer.
2. OpenShift routes that request to one of two Pods (Say, MyKieServer-Pod-X)
3. MyKieServer-Pod-X processes that Command, and at the end, it updates the KieServerState by calling KieServerStateRepository.store()
4. Then, the KieServerState ConfigMap will have a new '<container>' element. In addition, a DC roll-out will be triggered. 
5. The MyKieServer DC will terminate current Pods and create new Pods in a rolling fashion or all together depending on 'strategy: Recreate' or 'strategy: Rolling'. 
6. When the new Pods, MyKieServer-Pod-Alfa and MyKieServer-Pod-Beta, are booting up, they will have a single consistent view of KieServeState with newly deployed KieContainer.

Keep in mind, Pods are stateless, ephemeral and anonymous in general. A Pod local file system will be gone once it has been destroyed. Having said that. there are number ways to add statefulness, one of which is what has been implemented here.

Reply:
@rhtevan thanks for detailed reply.

my main question though is if we really want to support this case - cluster of KIE Servers that deployment to single instance of the kie server would cause rollout? Personally I would say we should not. If you want that then you should rely on controller or made the changes to config map directly.

Reply:
I think the main point here is whether to allow Kie servers to modify config map or not.

Reply:
@rhtevan @sutaakar if we think having this supported (Kie Server writing updates to config map after updates via Kie Server management api) than I am fine with that.